app/controllers/uploads_controller.rb

```diff
@@ -187,6 +192,81 @@ def metadata
     }
   end
 
+  def generate_presigned_put
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    file_name = params.require(:file_name)
+    type = params.require(:type)
+
+    RateLimiter.new(
+      current_user, "generate-presigned-put-upload-stub", PRESIGNED_PUT_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    # don't want people posting arbitrary S3 metadata so we just take the
+    # one we need. all of these will be converted to x-amz-meta- metadata
+    # fields in S3 so it's best to use dashes in the names for consistency
+    #
+    # this metadata is baked into the presigned url and is not altered when
+    # sending the PUT from the clientside to the presigned url
+    metadata = if params[:metadata].present?
+      meta = {}
+      if params[:metadata]["sha1-checksum"].present?
+        meta["sha1-checksum"] = params[:metadata]["sha1-checksum"]
+      end
+      meta
+    end
+
+    url = Discourse.store.signed_url_for_temporary_upload(
+      file_name, metadata: metadata
+    )
+    key = Discourse.store.path_from_url(url)
+
+    upload_stub = ExternalUploadStub.create!(
+      key: key,
+      created_by: current_user,
+      original_filename: file_name,
+      upload_type: type
+    )
+
+    render json: { method: :put, url: url, key: key, unique_identifier: upload_stub.unique_identifier }
```

Comment:
I'm curious about the `method: :put` here. It doesn't change so can we just assume it is always `PUT` in the uppy upload plugin on the client side?

Reply:
Yes good point; you can do POST presigned URL but that has a whole form in it and it's more complex; we will always using PUT so I will hardcode that.